# Αναζήτηση: BoW και TF-IDF/BM25 Retrieval -- Μέρος 1ο

<img src="https://drive.google.com/uc?export=view&id=1rRynSvwCZgEDPunOQT8wLqflmvVJz5w-" width="350">

### **Αναζήτηση Κειμένου**
Η αναζήτηση κειμένου είναι θεμελιώδες κομμάτι των σύγχρονων συστημάτων τεχνητής νοημοσύνης. Το βασικό πρόβλημα είναι η εύρεση ή ανάκτηση κειμένων από μια συλλογή κειμένων — τη λεγόμενη **knowledge base** — δεδομένου ενός **query**, για παράδειγμα μιας ερώτησης χρήστη. Με βάση αυτό το query, θέλουμε να ανακτήσουμε τα πιο σχετικά έγγραφα σε κάποια κατάταξη. Τα ανακτηθέντα έγγραφα μπορεί να είναι το τελικό αποτέλεσμα της διαδικασίας ή μπορεί να τροφοδοτούν ένα άλλο ML/AI component, παραδείγματος χάριν ένα γλωσσικό μοντέλο (LLM), σκοπός του οποίου είναι να απαντήσει στην ερώτηση του χρήστη, τεκμηριώνοντας την απάντηση με συγκεκριμένα έγγραφα από την knowledge base. Ο στόχος αυτού και του επόμενου notebook είναι να εξερευνήσουμε διαφορετικές προσεγγίσεις για την επίλυση αυτού του προβλήματος.

### Θα υλοποιήσουμε και θα συγκρίνουμε τρεις κύριες προσεγγίσεις:

1. **Bag of Words Search**: Η baseline μέθοδός μας, που ταιριάζει έγγραφα με βάση την απλή επικάλυψη λέξεων (μετράμε τον αριθμό κοινών λέξεων της ερώτησης με κάθε έγγραφο/κείμενο). Ενώ είναι διαισθητική, αυτή η προσέγγιση έχει σημαντικούς περιορισμούς:
   - Θεωρεί όλες τις λέξεις ισοδύναμες
   - Δεν λαμβάνει υπόψη τη σημαντικότητα των όρων
   - Κοιτάζει μόνο για ακριβή ταιριάσματα λέξεων και, ως συνέπεια, δεν μπορεί να χειριστεί συνώνυμα ή σχετικές έννοιες

2. **Sparse Retrieval (TF-IDF/BM25)**: Μια πιο εξελιγμένη προσέγγιση που:
   - Ζυγίζει τους όρους με βάση τη σημαντικότητά τους στα έγγραφα
   - Λαμβάνει υπόψη τη συχνότητα των όρων και την inverse document frequency
   - Χειρίζεται καλύτερα τις μεταβολές στο μήκος των εγγράφων
   - Χρησιμοποιεί inverted index για αποτελεσματική ανάκτηση
   - Επιτυγχάνει καλή επίδοση σε ακριβή ταιριάσματα λέξεων και χρησιμοποιείται ευρέως σε συστήματα παραγωγής

Στο επόμενο notebook, θα εξετάσουμε:

3. **Dense Retrieval**: Μια εναλλακτική και πιο σύγχρονη προσέγγιση που:
   - Χρησιμοποιεί neural networks για να μάθει semantic representations
   - Μπορεί να ταιριάξει έγγραφα με βάση το νόημα, όχι μόνο τις λέξεις
   - Χειρίζεται φυσικά συνώνυμα και σχετικές έννοιες
   - Απαιτεί περισσότερους υπολογιστικούς πόρους
   - Επιτυγχάνει παρόμοια επίδοση με το TF-IDF σε μετρικές αξιολόγησης (metrics), αλλά με διαφορετικά πλεονεκτήματα

## Πειράματα

Θα αξιολογήσουμε αυτές τις μεθόδους χρησιμοποιώντας ένα απλό και μικρό σώμα κειμένων που δημιουργούμε εμείς, και στη συνέχεια το dataset NFCorpus, που περιέχει άρθρα PubMed και queries σχετικά με τη διατροφή. Για κάθε μέθοδο, θα:

1. Υλοποιήσουμε το σύστημα αναζήτησης
2. Εκτελέσουμε αναζητήσεις σε ένα δείγμα queries
3. Αξιολογήσουμε την επίδοση χρησιμοποιώντας μετρικές αξιολόγησης (metrics):
   - NDCG (Normalized Discounted Cumulative Gain)
   - Recall σε διαφορετικά cutoffs (k=5, 10, 20, 100)
4. Συγκρίνουμε τα αποτελέσματα ποιοτικά μέσα από διάφορα queries
5. Αναλύσουμε τα πλεονεκτήματα και τους περιορισμούς κάθε προσέγγισης
6. Δείξουμε πώς το TF-IDF και το BoW χειρίζονται διαφορετικά τα queries

Αυτή η πορεία από απλές σε πιο εξελιγμένες μεθόδους θα βοηθήσει να επεξηγήσουμε τους συμβιβασμούς μεταξύ διαφορετικών τεχνικών ανάκτησης και πότε καθεμία μπορεί να είναι καταλληλότερη.

```
Κωνσταντίνος Καραμανής: constantine@utexas.edu
http://users.ece.utexas.edu/~cmcaram/
The University of Texas at Austin
Archimedes/Athena RC
```


### Εγκατάσταση Βιβλιοθηκών

Αυτό το notebook χρησιμοποιεί πολλές εξειδικευμένες βιβλιοθήκες για information retrieval και επεξεργασία κειμένου. Είναι αφορμή να αναπτύξουμε και εμείς μια επαφή με αυτές τις βιβλιοθήκες και τα εργαλεία που μας προσφέρουν.

- **NLTK** (Natural Language Toolkit): Μια περιεκτική πλατφόρμα για επεξεργασία φυσικής γλώσσας (natural language processing) στην Python. Τη χρησιμοποιούμε για εργασίες προεπεξεργασίας κειμένου, όπως tokenization και αφαίρεση stopwords.

- **Pyserini**: Μια Python διεπαφή στη βιβλιοθήκη αναζήτησης Lucene, που παρέχει εργαλεία για την κατασκευή και αναζήτηση text indexes. Είναι ιδιαίτερα χρήσιμη για sparse retrieval (αναζήτηση βασισμένη σε keywords).

- **BEIR** (Benchmarking IR): Ένα heterogeneous benchmark για Information Retrieval, που παρέχει σύνολα δεδομένων (standardized datasets) και evaluation metrics τόσο για sparse όσο και για dense retrieval μεθόδους.

Αυτές οι βιβλιοθήκες μάς επιτρέπουν να πειραματιστούμε και να συγκρίνουμε διαφορετικές προσεγγίσεις ανάκτησης πληροφορίας (information retrieval).

Σημείωση: Εάν σας ζητηθεί να επανεκκινήσετε το runtime μετά την εγκατάσταση αυτών των βιβλιοθηκών, κάντε κλικ στο "Ναι".


## Προσοχή

Η εισαγωγή των παρακάτω βιβλιοθηκών θέλει προσοχή. Ακολουθήστε αυτά τα βήματα:

1. Runtime --> Disconnect and delete runtime
2. Τρέξτε μόνο μια φορά το παρακάτω κελί
3. Runtime --> Restart session (όχι το "delete runtime")
4. **Μην** ξανατρέξετε το αμέσως επόμενο κελί, αλλά συνεχίστε πιο κάτω.

In [1]:
import os
import sys
import subprocess

def run(cmd):
    print("$", " ".join(cmd))
    subprocess.check_call(cmd)

run(["apt-get", "update", "-qq"])
run(["apt-get", "install", "-y", "-qq", "openjdk-21-jdk"])

java_home = "/usr/lib/jvm/java-21-openjdk-amd64"
os.environ["JAVA_HOME"] = java_home
os.environ["PATH"] = f"{java_home}/bin:" + os.environ["PATH"]

run([sys.executable, "-m", "pip", "install", "--upgrade", "pip"])

# This is the package that was dragging transformers into the broken PIL path.
run([sys.executable, "-m", "pip", "uninstall", "-y", "torchvision"])

# Only force-reinstall Pillow. Do NOT force-reinstall the whole ML stack.
run([
    sys.executable, "-m", "pip", "install",
    "--no-cache-dir", "--force-reinstall",
    "Pillow==11.3.0",
])

# Install the sparse-retrieval dependencies first.
run([
    sys.executable, "-m", "pip", "install",
    "--no-cache-dir",
    "pyserini==2.3.0",
    "faiss-cpu",
    "beir",
    "nltk",
])

print("Setup complete. Now go to Runtime -> Restart session.")

$ apt-get update -qq
$ apt-get install -y -qq openjdk-21-jdk
$ /usr/bin/python3 -m pip install --upgrade pip
$ /usr/bin/python3 -m pip uninstall -y torchvision
$ /usr/bin/python3 -m pip install --no-cache-dir --force-reinstall Pillow==11.3.0
$ /usr/bin/python3 -m pip install --no-cache-dir pyserini==2.3.0 faiss-cpu beir nltk
Setup complete. Now go to Runtime -> Restart session.


In [1]:
import os
import shutil
import subprocess

java_home = "/usr/lib/jvm/java-21-openjdk-amd64"
os.environ["JAVA_HOME"] = java_home
os.environ["PATH"] = f"{java_home}/bin:" + os.environ["PATH"]

print("java:", shutil.which("java"))
subprocess.check_call(["java", "-version"])

import PIL
print("Pillow:", PIL.__version__, PIL.__file__)

from pyserini.search.lucene import LuceneSearcher
print("LuceneSearcher import: ok")

java: /usr/lib/jvm/java-21-openjdk-amd64/bin/java
Pillow: 12.2.0 /usr/local/lib/python3.12/dist-packages/PIL/__init__.py
LuceneSearcher import: ok


In [2]:
from tqdm import tqdm
import json
import math
import tempfile

from beir import util
from beir.datasets.data_loader import GenericDataLoader

## Φορτώνουμε τις απαραίτητες βιβλιοθήκες.

### Core Imports
- **re**: Η ενσωματωμένη βιβλιοθήκη regular expressions της Python για pattern matching σε κείμενο
- **collections.Counter**: Αποτελεσματικό container για μέτρηση hashable objects
- **tqdm**: Utility progress bar για την παρακολούθηση μακροχρόνιων λειτουργιών

### NLTK Components
- **PorterStemmer**: Υλοποιεί τον αλγόριθμο stemming Porter για αναγωγή λέξεων στη ρίζα τους
- **word_tokenize**: Η συνάρτηση tokenization λέξεων της NLTK για διαίρεση κειμένου σε individual tokens

### NLTK Data Setup
Ο κώδικας κατεβάζει αυτόματα τους απαιτούμενους πόρους της NLTK, εάν δεν είναι ήδη διαθέσιμοι:
- `punkt`: Ένα pre-trained model για tokenization προτάσεων
- `punkt_tab`: Πρόσθετα δεδομένα tokenization για ειδικές περιπτώσεις

Αυτά τα components αποτελούν τη βάση της pipeline προεπεξεργασίας κειμένου μας, διασφαλίζοντας ότι έχουμε όλα τα απαραίτητα εργαλεία για tokenization και κανονικοποίηση κειμένου.


In [3]:
import re
from collections import Counter
from typing import Dict, List, Optional
import nltk
from nltk.stem import PorterStemmer
from nltk.tokenize import word_tokenize
import math
from tqdm import tqdm

# Download required NLTK data
try:
    nltk.data.find('tokenizers/punkt')
except LookupError:
    nltk.download('punkt')

try:
    nltk.data.find('tokenizers/punkt_tab')
except LookupError:
    nltk.download('punkt_tab')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


# Επεξήγηση Εργαλεία NLP


In [4]:
# Simple illustrations of the imported tools

# Example text to work with
sample_text = "The cats are running quickly. They're running faster than dogs!"

print("=" * 60)
print("1. Regular Expressions (re)")
print("=" * 60)
# Find all words that start with 'r'
words_starting_with_r = re.findall(r'\br\w+', sample_text, re.IGNORECASE)
print(f"Words starting with 'r': {words_starting_with_r}")

# Remove punctuation
text_no_punct = re.sub(r'[^\w\s]', '', sample_text)
print(f"Text without punctuation: {text_no_punct}")

1. Regular Expressions (re)
Words starting with 'r': ['running', 're', 'running']
Text without punctuation: The cats are running quickly Theyre running faster than dogs


In [5]:
print("\n" + "=" * 60)
print("2. NLTK Tokenization (word_tokenize)")
print("=" * 60)
# Tokenize the text into words
tokens = word_tokenize(sample_text.lower())
print(f"Tokenized text: {tokens}")




2. NLTK Tokenization (word_tokenize)
Tokenized text: ['the', 'cats', 'are', 'running', 'quickly', '.', 'they', "'re", 'running', 'faster', 'than', 'dogs', '!']


In [6]:
print("\n" + "=" * 60)
print("3. NLTK Stemming (PorterStemmer)")
print("=" * 60)
# Create a stemmer
stemmer = PorterStemmer()
# Stem each token
stemmed = [stemmer.stem(token) for token in tokens]
print(f"Original tokens: {tokens}")
print(f"Stemmed tokens: {stemmed}")





3. NLTK Stemming (PorterStemmer)
Original tokens: ['the', 'cats', 'are', 'running', 'quickly', '.', 'they', "'re", 'running', 'faster', 'than', 'dogs', '!']
Stemmed tokens: ['the', 'cat', 'are', 'run', 'quickli', '.', 'they', "'re", 'run', 'faster', 'than', 'dog', '!']


In [7]:
print("\n" + "=" * 60)
print("4. Counter (from collections)")
print("=" * 60)
# Count word frequencies
word_counts = Counter(tokens)
print(f"Word frequencies: {dict(word_counts)}")
print(f"Most common 3 words: {word_counts.most_common(3)}")




4. Counter (from collections)
Word frequencies: {'the': 1, 'cats': 1, 'are': 1, 'running': 2, 'quickly': 1, '.': 1, 'they': 1, "'re": 1, 'faster': 1, 'than': 1, 'dogs': 1, '!': 1}
Most common 3 words: [('running', 2), ('the', 1), ('cats', 1)]


In [8]:
print("\n" + "=" * 60)
print("5. Math (simple example)")
print("=" * 60)
# Calculate log for IDF-like calculation
num_docs = 100
docs_with_word = 5
idf = math.log(num_docs / docs_with_word)
print(f"IDF calculation: log({num_docs}/{docs_with_word}) = {idf:.3f}")


5. Math (simple example)
IDF calculation: log(100/5) = 2.996


## Μερικές Πρώτες Προσπάθειες

Θα δοκιμάσουμε μια φυσική ιδέα: το Bag-of-Words.

1. Έλεγχος για την παρουσία λέξεων από το query μας στα έγγραφα και βαθμολόγηση των εγγράφων αναλόγως.

Θα μπορούσαμε να προσπαθήσουμε να βελτιώσουμε το Bag-of-Words (αλλά δεν θα το κάνουμε σε αυτό το notebook):

2. Έλεγχος για την παρουσία bigrams ή trigrams λέξεων και βαθμολόγηση παρόμοια με την παραπάνω.
3. Μια πιο εξελιγμένη εκδοχή της πρώτης επιλογής, όπου λαμβάνουμε υπόψη τη συχνότητα των λέξεων.


## Υλοποίηση Bag of Words Search

Υλοποιούμε ένα απλό σύστημα αναζήτησης Bag-of-Words. Αυτό θα χρησιμεύσει ως σημείο σύγκρισης με πιο εξελιγμένες μεθόδους ανάκτησης: TF-IDF σε αυτό το notebook, και σημασιολογική αναζήτηση στο επόμενο.


### Core Classes

Πρώτα, υλοποιούμε ένα βασικό σύστημα information retrieval χρησιμοποιώντας το μοντέλο Bag of Words (BoW). Η κλάση `BagOfWordsSearch` παρέχει έναν απλό αλλά αποτελεσματικό τρόπο αναζήτησης σε έγγραφα με βάση επικάλυψη λέξεων.

### Key Components

#### Initialization
- Δημιουργεί ένα άδειο αποθετήριο εγγράφων
- Προαιρετική υποστήριξη stemming χρησιμοποιώντας Porter Stemmer (μειώνει τις λέξεις στη ρίζα τους)

#### Document Processing
- **Text Preprocessing** (`_preprocess_text`):
  - Μετατρέπει το κείμενο σε πεζά γράμματα
  - Αφαιρεί χαρακτήρες που δεν αντιστοιχούν σε γράμματα ή αριθμούς
  - Εφαρμόζει προαιρετικά stemming (π.χ., "running" → "run")

#### Search Functionality
- **Search Method**:
  - Παίρνει ένα query string και προαιρετική παράμετρο `top_k`
  - Υπολογίζει τη σχετικότητα εγγράφων χρησιμοποιώντας την επικάλυψη: set intersection
  - Score = αριθμός επικαλυπτόμενων λέξεων μεταξύ query και εγγράφου
  - Επιστρέφει έγγραφα καταταγμένα κατά relevance score



In [9]:
class BagOfWordsSearch:
    def __init__(self, use_stemming: bool = False):
        """Initialize the Bag of Words search system."""
        self.documents: Dict[str, str] = {}
        self.use_stemming = use_stemming
        if use_stemming:
            self.stemmer = PorterStemmer()

    def add_document(self, doc_id: str, text: str) -> None:
        """Add a document to the search index."""
        self.documents[doc_id] = text

    def _preprocess_text(self, text: str) -> List[str]:
        """Preprocess text by tokenizing and optionally stemming."""
        tokens = word_tokenize(text.lower())
        tokens = [re.sub(r'[^a-z0-9]', '', token) for token in tokens]
        tokens = [token for token in tokens if token]

        if self.use_stemming:
            tokens = [self.stemmer.stem(token) for token in tokens]

        return tokens

    def search(self, query: str, top_k: Optional[int] = None) -> List[tuple]:
        """Search documents using bag of words matching."""
        query_tokens = set(self._preprocess_text(query))
        results = []

        for doc_id, text in self.documents.items():
            doc_tokens = set(self._preprocess_text(text))
            score = len(query_tokens & doc_tokens) # '&' is the intersection operator
            results.append((doc_id, score))

        results.sort(key=lambda x: x[1], reverse=True)

        if top_k is not None:
            results = results[:top_k]

        return results

## Ένα Παράδειγμα

In [10]:
toy_corpus = [
"Clinic newsletter: patients with type 2 diabetes were reminded that medication, diet, and exercise work together; the article mentions metformin as a common start and notes that any new drug should prompt an interaction check during treatment review.",
"Follow-up note: the nurse reviewed side effects and asked about supplements, alcohol, and prescriptions, explaining that interaction screening matters even for familiar medicines like metformin when treatment plans change.",
"Educational summary: metformin is often first-line; the piece repeats that clinicians monitor kidney function and ask about interaction risks, because dehydration or other drugs can alter metformin safety in routine treatment.",
"Research digest: a small study of patients on metformin discussed adherence and lifestyle; it briefly mentions interaction concerns but focuses on diet, sleep, and long-term treatment goals.",
"Discharge instructions: continue metformin with meals, record glucose, and bring a medication list; the sheet says ‘tell us about any interaction with new pills’ as part of safe treatment planning.",
"Review paragraph: metformin lowers hepatic glucose output; the author adds that clinicians consider renal function and interaction flags, especially when treatment includes multiple agents.",
"Pharmacy handout: before refills, ask about interaction questions; it lists common warnings (alcohol, dehydration, contrast dye) and says metformin is usually well tolerated when treatment is stable.",
"Patient story: after starting metformin, the patient improved A1C; the narrative mentions an interaction screening at each visit but is mostly about motivation, meal planning, and staying on treatment.",
"Guideline excerpt: reassess treatment every few months; check labs, reinforce activity, and review interaction lists; metformin remains a baseline drug for many patients.",
"Clinician memo: lactic acidosis is rare; avoid metformin during acute illness; the memo repeats that interaction review is routine but does not describe any mechanism.",
"FAQ page: how metformin works, how to take metformin, and why metformin continues; it includes a generic line about interaction checks during treatment but no examples.",
"Community brochure: programs support patients with reminders, coaching, and medication literacy; it encourages asking pharmacists about interaction risks when taking metformin and other drugs.",
"Case narrative: missed appointments complicated treatment; the clinician emphasized consistent metformin dosing and repeated interaction checks as new prescriptions were added over time.",
"Imaging note: pause metformin around iodinated contrast; the note frames this as an interaction-adjacent precaution tied to kidney stress and follow-up testing after treatment.",
"Lecture transcript: ‘interaction’ is a broad pharmacology term; the speaker mentions metformin in passing and stresses that interaction questions should be documented for patient safety.",
"Pharmacokinetic note (the one we want): cimetidine reduces renal tubular secretion, increasing plasma levels; monitoring recommended when co-administered.",
"Abstract: the study compared two treatment groups; it uses the word interaction once and metformin once, but otherwise stays generic about outcomes in patients.",
"Clinic checklist: confirm metformin, confirm allergies, and ask about interaction with any new medication; the checklist is repetitive but not specific about which drug interacts.",
"Call center script: if patients report a new prescription, staff log a possible interaction and route the metformin question to pharmacy, but the script does not mention cimetidine or any mechanism.",
"Medication safety guide: always check for metformin interaction before prescribing new drugs; metformin interaction risks must be assessed; patients on metformin need interaction screening; metformin interaction protocols should be followed; never overlook potential metformin interaction.",
"Treatment protocol document: standard procedure involves checking metformin interaction with all concurrent medications; metformin interaction databases should be consulted; any metformin interaction concerns require immediate review; metformin interaction warnings are critical for patient safety.",
"Clinical reminder bulletin: metformin interaction metformin interaction metformin interaction—this bulletin emphasizes metformin interaction monitoring; staff must document every metformin interaction check; patient records must show metformin interaction assessment; all metformin interaction reports require follow-up.",
"Hospital policy manual: section on metformin interaction management; metformin interaction screening is mandatory; metformin interaction documentation requirements; metformin interaction reporting procedures; metformin interaction training modules; comprehensive metformin interaction guidelines.",
"Pharmacy training manual: metformin interaction protocols; metformin interaction safety checks; metformin interaction review process; metformin interaction documentation standards; metformin interaction assessment tools; metformin interaction guidelines for pharmacists.",
"Quality assurance checklist: verify metformin interaction screening completed; confirm metformin interaction documentation present; ensure metformin interaction alerts reviewed; check metformin interaction flags addressed; validate metformin interaction protocols followed.",
"Standard operating procedure: step one check metformin interaction; step two document metformin interaction; step three review metformin interaction; step four report metformin interaction; all metformin interaction steps mandatory."
]
toy_query = "I want to learn about cimetidine metformin interaction"


## Ένα απλό παράδειγμα του BoW μας

In [11]:
doc_id = 3  # which document to add to the index -- start at 0
doc_text = toy_corpus[doc_id]  # Use as integer index

bow_search = BagOfWordsSearch(use_stemming=True)
bow_search.add_document(str(doc_id), doc_text)  # Convert to string when adding

results = bow_search.search(toy_query)
print(f"Document {results[0][0]} has score (word overlap) {results[0][1]:.1f}")

Document 3 has score (word overlap) 2.0


In [12]:
# Now let's add all documents to the index and search for the query

for i in range(len(toy_corpus)):
    doc_id = i
    doc_text = toy_corpus[i]
    bow_search.add_document(str(doc_id), doc_text)

results = bow_search.search(toy_query)

# sort results by score
results.sort(key=lambda x: x[1], reverse=True)

# print the top 10 results
for i in range(len(results)):
    print(f"Document {results[i][0]} has score (word overlap) {results[i][1]:.1f}")


Document 10 has score (word overlap) 4.0
Document 18 has score (word overlap) 4.0
Document 1 has score (word overlap) 3.0
Document 2 has score (word overlap) 3.0
Document 4 has score (word overlap) 3.0
Document 6 has score (word overlap) 3.0
Document 7 has score (word overlap) 3.0
Document 11 has score (word overlap) 3.0
Document 16 has score (word overlap) 3.0
Document 17 has score (word overlap) 3.0
Document 3 has score (word overlap) 2.0
Document 0 has score (word overlap) 2.0
Document 5 has score (word overlap) 2.0
Document 8 has score (word overlap) 2.0
Document 9 has score (word overlap) 2.0
Document 12 has score (word overlap) 2.0
Document 13 has score (word overlap) 2.0
Document 14 has score (word overlap) 2.0
Document 15 has score (word overlap) 2.0
Document 19 has score (word overlap) 2.0
Document 20 has score (word overlap) 2.0
Document 21 has score (word overlap) 2.0
Document 22 has score (word overlap) 2.0
Document 23 has score (word overlap) 2.0
Document 24 has score (wor

### Iterators

Ας χρησιμοποιήσουμε iterators, ώστε να αξιοποιήσουμε σωστά τον τρόπο με τον οποίο η Python χειρίζεται τις λίστες.


In [13]:
# Create a BagOfWordsSearch instance
bow_search = BagOfWordsSearch(use_stemming=True)

# Add all documents from toy_corpus
print("Adding documents to the search index...")
for i, doc_text in enumerate(toy_corpus):
    doc_id = f"doc_{i}"
    bow_search.add_document(doc_id, doc_text)

print(f"Added {len(toy_corpus)} documents\n")

# Process the query to show what tokens are extracted
print("Query tokens after preprocessing:")
query_tokens = set(bow_search._preprocess_text(toy_query))
print(query_tokens)
print()

# Search for documents matching the query
print(f"Searching for: '{toy_query}'\n")
results = bow_search.search(toy_query, top_k=10)

# Display results with detailed overlap information
print("Top 10 Results (showing overlap details):")
print("=" * 80)
for rank, (doc_id, score) in enumerate(results, 1):
    doc_idx = int(doc_id.split('_')[1])
    doc_text = toy_corpus[doc_idx]

    # Show which tokens overlap
    doc_tokens = set(bow_search._preprocess_text(doc_text))
    overlapping_tokens = query_tokens & doc_tokens

    print(f"\nRank {rank}: Score = {score} (overlap of {len(overlapping_tokens)} tokens)")
    print(f"Overlapping tokens: {overlapping_tokens}")
    print(f"Document preview: {doc_text[:100]}...")
    print("-" * 80)

Adding documents to the search index...
Added 26 documents

Query tokens after preprocessing:
{'about', 'want', 'to', 'learn', 'cimetidin', 'i', 'metformin', 'interact'}

Searching for: 'I want to learn about cimetidine metformin interaction'

Top 10 Results (showing overlap details):

Rank 1: Score = 4 (overlap of 4 tokens)
Overlapping tokens: {'to', 'interact', 'metformin', 'about'}
Document preview: FAQ page: how metformin works, how to take metformin, and why metformin continues; it includes a gen...
--------------------------------------------------------------------------------

Rank 2: Score = 4 (overlap of 4 tokens)
Overlapping tokens: {'to', 'cimetidin', 'metformin', 'interact'}
Document preview: Call center script: if patients report a new prescription, staff log a possible interaction and rout...
--------------------------------------------------------------------------------

Rank 3: Score = 3 (overlap of 3 tokens)
Overlapping tokens: {'interact', 'metformin', 'about'}
Docu

## Συζήτηση: Προβλήματα του BoW

Το πιο σχετικό αποτέλεσμα δεν φαίνεται να εμφανίζεται στη λίστα.
Πράγματι, το 16ο έγγραφο (doc 15, αρίθμηση από 0) είναι το πιο σχετικό έγγραφο:
``
"Pharmacokinetic note (αυτό που θέλουμε): cimetidine reduces renal tubular secretion, increasing plasma levels; monitoring recommended when co-administered.",
``
Ωστόσο, έχει επικάλυψη (BoW score) μόνο 2, και ως εκ τούτου κατατάσσεται πολύ χαμηλά στην BoW κατάταξη.

Ποια είναι η αιτία;

* Άλλα έγγραφα έχουν μεγαλύτερη επικάλυψη λόγω λέξεων που δεν είναι σημαντικές, όπως "about", "interact" ή "to".  
* Η λέξη "cimetidin" είναι η κρίσιμη λέξη στο query, αλλά δεν αρκεί για να ενισχύσει το σωστό έγγραφο στην κατάταξη, διότι δεν έχουμε (αυτόματο) τρόπο να εκτιμήσουμε τη σημασία (αξία) της λέξης για την συγκεκριμένη αναζήτηση.

Στη συνέχεια περιγράφουμε μια προσέγγιση που προσπαθεί να διορθώσει αυτήν την αδυναμία του BoW. Οι βασικές ιδέες είναι:
* TF -- Term Frequency: εάν μια λέξη του query έχει υψηλότερη συχνότητα στο έγγραφο, αυτό θα πρέπει να ζυγίζεται θετικά.
* IDF -- Inverse Document Frequency: εάν μια λέξη του query εμφανίζεται σε πολλά έγγραφα, τότε είναι λιγότερο σημαντική/διακριτική, και θα πρέπει να ζυγίζεται αρνητικά.



## Sparse Search: TF-IDF

Το TF-IDF μετρά πόσο σημαντική είναι μια λέξη σε ένα έγγραφο σε μια συλλογή. Συνδυάζει δύο παράγοντες:
1. Πόσο συχνά εμφανίζεται η λέξη στο έγγραφο (TF)
2. Πόσο μοναδική είναι η λέξη σε όλα τα έγγραφα (IDF)

### Term Frequency (TF)
- Μετρά πόσο συχνά εμφανίζεται μια λέξη σε ένα έγγραφο
- Περισσότερες εμφανίσεις → πιο σημαντική
```python
tf = (αριθμός φορών που εμφανίζεται ο όρος στο έγγραφο) / (συνολικές λέξεις στο έγγραφο)
```

### Inverse Document Frequency (IDF)
- Μετρά πόσο μοναδική ή σπάνια είναι μια λέξη σε όλα τα έγγραφα
- Οι σπάνιες λέξεις παίρνουν υψηλότερα scores, οι κοινές λέξεις παίρνουν χαμηλότερα scores
```python
idf = log(συνολικός αριθμός εγγράφων / αριθμός εγγράφων που περιέχουν τον όρο)
```

### Συνδυασμός
- TF-IDF = TF × IDF
- Υψηλό score σημαίνει:
  - Η λέξη εμφανίζεται συχνά σε αυτό το έγγραφο (υψηλό TF)
  - Η λέξη εμφανίζεται σε λίγα άλλα έγγραφα (υψηλό IDF)

### Παράδειγμα
Σκεφτείτε τη λέξη "neural" vs "transformer" σε ερευνητικά papers για νευρωνικά δίκτυα:
- "neural":
  - Υψηλό TF (εμφανίζεται συχνά)
  - Χαμηλό IDF (εμφανίζεται σε κάθε έγγραφο)
  - Χαμηλό TF-IDF score
- "transformer":
  - Μέτριο TF (μπορεί να εμφανίζεται αρκετές φορές)
  - Υψηλό IDF (εμφανίζεται σε λιγότερα έγγραφα)
  - Υψηλότερο TF-IDF score

Αυτός είναι ο λόγος που το TF-IDF είναι καλύτερο από την απλή μέτρηση λέξεων (επικάλυψη) - χειρίζεται αυτόματα τη σημαντικότητα των λέξεων με βάση τη μοναδικότητά τους στο συνολικό σώμα κειμένων (text corpus).




## Σύγκριση Bag-of-Words και TF-IDF

In [14]:
# First, let's run Bag of Words search on the toy corpus
print("=" * 80)
print("Bag of Words (BoW) Search Results")
print("=" * 80)

# Initialize BoW search
bow_search = BagOfWordsSearch(use_stemming=False)

# Add all documents to the index
for i, doc_text in enumerate(toy_corpus):
    bow_search.add_document(f"doc_{i}", doc_text)

# Run the search
print(f"\nQuery: '{toy_query}'")
print(f"\nTop 10 BoW Results (by word overlap count):")
print("-" * 80)

bow_results = bow_search.search(toy_query, top_k=10)
for rank, (doc_id, score) in enumerate(bow_results, 1):
    doc_idx = int(doc_id.split('_')[1])
    doc_preview = toy_corpus[doc_idx][:80] + "..." if len(toy_corpus[doc_idx]) > 80 else toy_corpus[doc_idx]
    print(f"{rank}. [Score: {score}] Doc {doc_idx}: {doc_preview}")

# Highlight the actual relevant document (find which document contains cimetidine)
cimetidine_doc_idx = None
for i, doc in enumerate(toy_corpus):
    if "cimetidine" in doc.lower():
        cimetidine_doc_idx = i
        break

if cimetidine_doc_idx is not None:
    rank_in_results = [i+1 for i, (did, _) in enumerate(bow_results) if int(did.split('_')[1]) == cimetidine_doc_idx]
    rank_text = str(rank_in_results[0]) if rank_in_results else 'NOT in top 10'
    print(f"\n⚠️  Note: The actual relevant document (Doc {cimetidine_doc_idx}) about cimetidine-metformin interaction")
    print(f"   appears at rank {rank_text}")
else:
    print(f"\n⚠️  Note: Could not find document containing 'cimetidine'")

print("\n" + "=" * 80)
print("TF-IDF Search Results (using sklearn)")
print("=" * 80)

# Now let's run TF-IDF using sklearn
from sklearn.feature_extraction.text import TfidfVectorizer
import numpy as np

# Initialize TF-IDF vectorizer
tfidf_vectorizer = TfidfVectorizer(
    lowercase=True,
    token_pattern=r'\b[a-z]+\b',  # Simple word tokenization
    max_features=None,
    norm='l2'
)

# Fit on corpus and transform
tfidf_matrix = tfidf_vectorizer.fit_transform(toy_corpus)

# Transform the query
query_vector = tfidf_vectorizer.transform([toy_query])

# Calculate cosine similarity
from sklearn.metrics.pairwise import cosine_similarity
similarities = cosine_similarity(query_vector, tfidf_matrix)[0]

# Get top results
top_indices = np.argsort(similarities)[::-1][:10]

print(f"\nQuery: '{toy_query}'")
print(f"\nTop 10 TF-IDF Results (by cosine similarity):")
print("-" * 80)

for rank, doc_idx in enumerate(top_indices, 1):
    similarity_score = similarities[doc_idx]
    doc_preview = toy_corpus[doc_idx][:80] + "..." if len(toy_corpus[doc_idx]) > 80 else toy_corpus[doc_idx]
    print(f"{rank}. [Score: {similarity_score:.4f}] Doc {doc_idx}: {doc_preview}")

# Highlight the actual relevant document
# Find which document actually contains "cimetidine" (use the one we found earlier, or find it)
if cimetidine_doc_idx is None:
    for i, doc in enumerate(toy_corpus):
        if "cimetidine" in doc.lower():
            cimetidine_doc_idx = i
            break

print(f"\n✓ The actual relevant document about cimetidine-metformin interaction")
if cimetidine_doc_idx is not None:
    # Calculate rank from all similarities
    all_ranks = np.argsort(similarities)[::-1]
    doc_rank = np.where(all_ranks == cimetidine_doc_idx)[0][0] + 1
    print(f"  Document (Doc {cimetidine_doc_idx}) appears at rank {doc_rank}")
    if doc_rank <= 10:
        print(f"  ✓ It's in the top 10! (TF-IDF successfully found the relevant document)")
    else:
        print(f"  ⚠️  It's not in the top 10, but still better ranked than many non-relevant docs")
else:
    print(f"  ⚠️  Could not find document containing 'cimetidine'")

print("\n" + "=" * 80)
print("Key Observations:")
print("=" * 80)
print("""
BoW Pitfalls Demonstrated:
1. BoW treats all words equally - documents mentioning 'metformin' and 'interaction'
   frequently but NOT 'cimetidine' can rank higher than the relevant one
2. No term weighting - common words like 'metformin' and 'interaction' dominate
3. The relevant document that specifically mentions 'cimetidine' may not
   rank at the top because it doesn't repeat 'metformin' and 'interaction' as often

TF-IDF Advantages:
1. Rare terms like 'cimetidine' get higher weight (high IDF)
2. Term frequency is normalized by document length
3. Documents matching rare query terms (like 'cimetidine') rank higher
4. Better at finding the specific document about cimetidine-metformin interaction
""")

Bag of Words (BoW) Search Results

Query: 'I want to learn about cimetidine metformin interaction'

Top 10 BoW Results (by word overlap count):
--------------------------------------------------------------------------------
1. [Score: 4] Doc 10: FAQ page: how metformin works, how to take metformin, and why metformin continue...
2. [Score: 4] Doc 18: Call center script: if patients report a new prescription, staff log a possible ...
3. [Score: 3] Doc 1: Follow-up note: the nurse reviewed side effects and asked about supplements, alc...
4. [Score: 3] Doc 2: Educational summary: metformin is often first-line; the piece repeats that clini...
5. [Score: 3] Doc 4: Discharge instructions: continue metformin with meals, record glucose, and bring...
6. [Score: 3] Doc 6: Pharmacy handout: before refills, ask about interaction questions; it lists comm...
7. [Score: 3] Doc 7: Patient story: after starting metformin, the patient improved A1C; the narrative...
8. [Score: 3] Doc 11: Community brochu

### TF-IDF: βήμα προς βήμα

Γιατί πετυχαίνει το TF-IDF; Ας υπολογίσουμε ρητά το TF-IDF score του.


In [15]:
import re
import math
from collections import Counter
from nltk.tokenize import word_tokenize

def preprocess_text(text):
    """Preprocess text: lowercase, tokenize, remove non-alphanumeric."""
    tokens = word_tokenize(text.lower())
    tokens = [re.sub(r'[^a-z0-9]', '', token) for token in tokens]
    tokens = [token for token in tokens if token]
    return tokens

# Select documents to analyze
doc_correct = toy_corpus[15]  # The correct document (Doc 16 in 1-based)
doc_high_overlap_1 = toy_corpus[10]  # Doc 11 - high word overlap
doc_high_overlap_2 = toy_corpus[18]  # Doc 19 - high word overlap

# Preprocess query and documents
query_tokens = preprocess_text(toy_query)
doc_correct_tokens = preprocess_text(doc_correct)
doc_high1_tokens = preprocess_text(doc_high_overlap_1)
doc_high2_tokens = preprocess_text(doc_high_overlap_2)

# Preprocess all documents to compute IDF
all_doc_tokens = [preprocess_text(doc) for doc in toy_corpus]
total_docs = len(toy_corpus)

print("=" * 80)
print("TF-IDF Calculation - Step by Step")
print("=" * 80)
print(f"\nQuery: '{toy_query}'")
print(f"Query tokens: {query_tokens}")
print(f"\nDocuments to analyze:")
print(f"  Doc 16 (correct): {doc_correct[:80]}...")
print(f"  Doc 11 (high overlap): {doc_high_overlap_1[:80]}...")
print(f"  Doc 19 (high overlap): {doc_high_overlap_2[:80]}...")

# Step 1: Calculate IDF for each query term across all documents
print("\n" + "=" * 80)
print("Step 1: Calculate IDF (Inverse Document Frequency)")
print("=" * 80)
print("\nIDF = log(total_documents / documents_containing_term)")
print(f"Total documents: {total_docs}\n")

idf_scores = {}
for term in query_tokens:
    # Count how many documents contain this term
    docs_with_term = sum(1 for doc_tokens in all_doc_tokens if term in doc_tokens)
    if docs_with_term > 0:
        idf = math.log(total_docs / docs_with_term)
        idf_scores[term] = idf
        print(f"  '{term}': appears in {docs_with_term}/{total_docs} docs → IDF = log({total_docs}/{docs_with_term}) = {idf:.4f}")
    else:
        idf_scores[term] = 0
        print(f"  '{term}': appears in 0/{total_docs} docs → IDF = 0 (term not in corpus)")

# Step 2: Calculate TF for each document and query term
print("\n" + "=" * 80)
print("Step 2: Calculate TF (Term Frequency) for each document")
print("=" * 80)
print("\nTF = (count of term in document) / (total terms in document)\n")

def calculate_tf(doc_tokens, term):
    """Calculate term frequency in document."""
    term_count = doc_tokens.count(term)
    total_terms = len(doc_tokens)
    if total_terms > 0:
        return term_count / total_terms, term_count, total_terms
    return 0, 0, 0

# Calculate TF for each document
print("Doc 16 (correct):")
doc_correct_tf = {}
for term in query_tokens:
    tf, count, total = calculate_tf(doc_correct_tokens, term)
    doc_correct_tf[term] = tf
    print(f"  '{term}': appears {count} times in {total} total terms → TF = {count}/{total} = {tf:.4f}")

print("\nDoc 11 (high overlap):")
doc_high1_tf = {}
for term in query_tokens:
    tf, count, total = calculate_tf(doc_high1_tokens, term)
    doc_high1_tf[term] = tf
    print(f"  '{term}': appears {count} times in {total} total terms → TF = {count}/{total} = {tf:.4f}")

print("\nDoc 19 (high overlap):")
doc_high2_tf = {}
for term in query_tokens:
    tf, count, total = calculate_tf(doc_high2_tokens, term)
    doc_high2_tf[term] = tf
    print(f"  '{term}': appears {count} times in {total} total terms → TF = {count}/{total} = {tf:.4f}")

# Step 3: Calculate TF-IDF for each document
print("\n" + "=" * 80)
print("Step 3: Calculate TF-IDF = TF × IDF for each query term")
print("=" * 80)

def calculate_tfidf(tf_dict, idf_dict):
    """Calculate TF-IDF scores for all query terms."""
    tfidf_dict = {}
    total_tfidf = 0
    for term in query_tokens:
        tf = tf_dict.get(term, 0)
        idf = idf_dict.get(term, 0)
        tfidf = tf * idf
        tfidf_dict[term] = tfidf
        total_tfidf += tfidf
    return tfidf_dict, total_tfidf

doc_correct_tfidf, doc_correct_score = calculate_tfidf(doc_correct_tf, idf_scores)
doc_high1_tfidf, doc_high1_score = calculate_tfidf(doc_high1_tf, idf_scores)
doc_high2_tfidf, doc_high2_score = calculate_tfidf(doc_high2_tf, idf_scores)

print("\nDoc 16 (correct):")
for term in query_tokens:
    tf = doc_correct_tf[term]
    idf = idf_scores.get(term, 0)
    tfidf = doc_correct_tfidf[term]
    print(f"  '{term}': TF({tf:.4f}) × IDF({idf:.4f}) = {tfidf:.4f}")
print(f"  Total TF-IDF score: {doc_correct_score:.4f}")

print("\nDoc 11 (high overlap):")
for term in query_tokens:
    tf = doc_high1_tf[term]
    idf = idf_scores.get(term, 0)
    tfidf = doc_high1_tfidf[term]
    print(f"  '{term}': TF({tf:.4f}) × IDF({idf:.4f}) = {tfidf:.4f}")
print(f"  Total TF-IDF score: {doc_high1_score:.4f}")

print("\nDoc 19 (high overlap):")
for term in query_tokens:
    tf = doc_high2_tf[term]
    idf = idf_scores.get(term, 0)
    tfidf = doc_high2_tfidf[term]
    print(f"  '{term}': TF({tf:.4f}) × IDF({idf:.4f}) = {tfidf:.4f}")
print(f"  Total TF-IDF score: {doc_high2_score:.4f}")

# Step 4: Summary and comparison
print("\n" + "=" * 80)
print("Step 4: Summary and Key Insights")
print("=" * 80)
print("\nFinal TF-IDF Scores (sum of TF-IDF for all query terms):")
print(f"  Doc 16 (correct):      {doc_correct_score:.4f}")
print(f"  Doc 11 (high overlap): {doc_high1_score:.4f}")
print(f"  Doc 19 (high overlap): {doc_high2_score:.4f}")

print("\n🔍 Key Observations:")
if 'cimetidine' in idf_scores:
    print(f"  1. 'cimetidine' has the highest IDF ({idf_scores.get('cimetidine', 0):.4f}) - it's the rarest term!")
print(f"  2. Doc 16 contains 'cimetidine', giving it a significant boost despite lower TF for common terms")
print(f"  3. Docs 11 and 19 have high TF for 'metformin' and 'interaction', but these terms have lower IDF")
print(f"  4. The rare term 'cimetidine' outweighs the common terms in TF-IDF scoring")
print("\n✅ This explains why TF-IDF ranks Doc 16 higher than BoW!")

TF-IDF Calculation - Step by Step

Query: 'I want to learn about cimetidine metformin interaction'
Query tokens: ['i', 'want', 'to', 'learn', 'about', 'cimetidine', 'metformin', 'interaction']

Documents to analyze:
  Doc 16 (correct): Pharmacokinetic note (the one we want): cimetidine reduces renal tubular secreti...
  Doc 11 (high overlap): FAQ page: how metformin works, how to take metformin, and why metformin continue...
  Doc 19 (high overlap): Call center script: if patients report a new prescription, staff log a possible ...

Step 1: Calculate IDF (Inverse Document Frequency)

IDF = log(total_documents / documents_containing_term)
Total documents: 26

  'i': appears in 0/26 docs → IDF = 0 (term not in corpus)
  'want': appears in 1/26 docs → IDF = log(26/1) = 3.2581
  'to': appears in 3/26 docs → IDF = log(26/3) = 2.1595
  'learn': appears in 0/26 docs → IDF = 0 (term not in corpus)
  'about': appears in 9/26 docs → IDF = log(26/9) = 1.0609
  'cimetidine': appears in 2/26 docs →

# Αναζήτηση: BoW και TF-IDF/BM25 Retrieval -- Μέρος 2ο

Τώρα περνάμε στο μεγαλύτερο πρόβλημα, με το σώμα κειμένων NFCorpus, το οποίο κατεβάζουμε από το benchmark [BeIR](https://arxiv.org/abs/2104.08663).


## Φόρτωση του Dataset: Dictionaries

Ας φορτώσουμε τώρα τα δεδομένα του dataset NFCorpus.

Το dataset περιέχει:

* Άρθρα PubMed (documents/corpus) — περίπου 3600
* Ερωτήσεις σχετικά με την υγεία και τη διατροφή (queries) — περίπου 300
* Ground truth (qrels): το "χρυσό κλειδί" που μας λέει πόσο σχετικό είναι το κάθε άρθρο (έγγραφο) με την κάθε ερώτηση (query). Αυτό μας επιτρέπει να υπολογίσουμε μετρικές αξιολόγησης των μεθόδων μας.

Το ζητούμενο είναι να ανακτήσουμε σχετικά άρθρα PubMed, δοθέντος ενός query.


In [16]:
out_dir = os.path.join(os.getcwd(), "datasets")

dataset = "nfcorpus"
url = "https://public.ukp.informatik.tu-darmstadt.de/thakur/BEIR/datasets/{}.zip".format(dataset)
data_path = util.download_and_unzip(url, out_dir)

corpus, queries, qrels = GenericDataLoader(data_folder=data_path).load(split="test")

/content/datasets/nfcorpus.zip:   0%|          | 0.00/2.34M [00:00<?, ?iB/s]

  0%|          | 0/3633 [00:00<?, ?it/s]

### Λεξικό: Κλειδιά και Τιμές (Dictionary: Keys και Values)

Έχουμε φορτώσει τρία λεξικά (dictionaries).
* Το ``corpus`` είναι ένα λεξικό εγγράφων
* Το ``queries`` είναι ένα λεξικό με queries
* Το ``qrels`` είναι ένα λεξικό με αξιολογήσεις σχετικότητας (relevance judgments). Αυτή είναι η "χρυσή απάντηση" (ground truth) μας.

Το ``corpus`` είναι ένα λεξικό στη μορφή `{doc_id: document}`.

Το ``corpus.keys()`` δίνει μια iterable όψη των κλειδιών του λεξικού, την οποία μπορούμε να μετατρέψουμε σε λίστα μέσω ``list(corpus.keys())``.
Το ``corpus.values()`` είναι μια όψη των τιμών του λεξικού.

Το ``queries`` είναι ένα λεξικό στη μορφή `{query_id: query}`.

Το ``qrels`` είναι ένα λεξικό στη μορφή `{query_id: [doc_id1: score1, doc_id2: score2, ...]}`. Εδώ, το `score` αντιπροσωπεύει πόσο σχετικό είναι το κάθε έγγραφο με το συγκεκριμένο query. Εάν `score3 > score1`, τότε το 3ο έγγραφο έχει μεγαλύτερη σχέση με το query απ' ό,τι το 1ο.

**Ας δούμε μερικά από τα δεδομένα**


In [17]:
# Get first document value
first_doc = list(corpus.values())[0]
print("This is the first document: ", first_doc)  # {'text': '...', 'title': '...'}

# Or get by key
first_key = list(corpus.keys())[0]
print("This is the first document key: ", first_key)
first_doc = corpus[first_key]  # Same result
print("This is the first document: ", first_doc)



This is the first document:  {'text': 'Recent studies have suggested that statins, an established drug group in the prevention of cardiovascular mortality, could delay or prevent breast cancer recurrence but the effect on disease-specific mortality remains unclear. We evaluated risk of breast cancer death among statin users in a population-based cohort of breast cancer patients. The study cohort included all newly diagnosed breast cancer patients in Finland during 1995–2003 (31,236 cases), identified from the Finnish Cancer Registry. Information on statin use before and after the diagnosis was obtained from a national prescription database. We used the Cox proportional hazards regression method to estimate mortality among statin users with statin use as time-dependent variable. A total of 4,151 participants had used statins. During the median follow-up of 3.25 years after the diagnosis (range 0.08–9.0 years) 6,011 participants died, of which 3,619 (60.2%) was due to breast cancer. Afte

In [18]:
# Get all keys
all_keys = list(corpus.keys())
print("This is the list of all document keys: ", all_keys)

# The first five keys in queries and a sample entry
print("This is the list of the first five query keys: ", list(queries.keys())[:5])
first_query_key = list(queries.keys())[0]  # Get a query ID, not a document ID
print("This is the first query: ", queries[first_query_key])  # Use the query ID

# Same for qrels
print("These are the first five query keys: ", list(qrels.keys())[:5])
print("These are the relevance judgements for the first query: ", qrels[first_query_key])  # Use the query ID here too
print("This is the text of one of the most relevant documents for the first query: ", corpus[list(qrels[first_query_key].keys())[0]])

This is the list of all document keys:  ['MED-10', 'MED-14', 'MED-118', 'MED-301', 'MED-306', 'MED-329', 'MED-330', 'MED-332', 'MED-334', 'MED-335', 'MED-398', 'MED-557', 'MED-666', 'MED-691', 'MED-692', 'MED-702', 'MED-706', 'MED-707', 'MED-708', 'MED-709', 'MED-711', 'MED-712', 'MED-713', 'MED-714', 'MED-716', 'MED-717', 'MED-718', 'MED-719', 'MED-720', 'MED-721', 'MED-722', 'MED-723', 'MED-724', 'MED-726', 'MED-727', 'MED-728', 'MED-729', 'MED-730', 'MED-731', 'MED-732', 'MED-733', 'MED-734', 'MED-735', 'MED-736', 'MED-743', 'MED-744', 'MED-745', 'MED-746', 'MED-748', 'MED-749', 'MED-751', 'MED-752', 'MED-753', 'MED-754', 'MED-756', 'MED-757', 'MED-758', 'MED-759', 'MED-760', 'MED-761', 'MED-762', 'MED-816', 'MED-818', 'MED-819', 'MED-820', 'MED-821', 'MED-822', 'MED-823', 'MED-824', 'MED-825', 'MED-826', 'MED-827', 'MED-828', 'MED-829', 'MED-830', 'MED-831', 'MED-832', 'MED-833', 'MED-834', 'MED-835', 'MED-836', 'MED-837', 'MED-838', 'MED-839', 'MED-840', 'MED-841', 'MED-842', 'MED

### Bag-of-Words και το NFCorpus Dataset

Έχουμε δει πώς λειτουργούν το Bag of Words και το TF-IDF στο απλό παράδειγμά μας.

**Μόνο 10 queries**: αρχικά, χρησιμοποιούμε μόνο τα πρώτα 10 queries!


In [19]:
# Configuration
sample_size = 10 # μόνο 10 queries
k_values = [5, 10, 20, 100]

# Prepare sample data
print("Preparing sample data...")
sample_query_ids = list(queries.keys())[:sample_size]
sample_queries = {qid: queries[qid] for qid in sample_query_ids}
sample_qrels = {qid: qrels[qid] for qid in sample_query_ids if qid in qrels}

# Initialize and run search
print("\nRunning bag of words search...")
bow_search = BagOfWordsSearch(use_stemming=True)

# Add documents
print("Adding documents to index...")
for doc_id, doc in tqdm(corpus.items(), desc="Indexing"):
    full_text = f"{doc['title']} {doc['text']}"
    bow_search.add_document(doc_id, full_text)

# Run searches
bow_results = {}
for query_id, query_text in tqdm(sample_queries.items(), desc="Searching"):
    results = bow_search.search(query_text, top_k=max(k_values))
    bow_results[query_id] = results

Preparing sample data...

Running bag of words search...
Adding documents to index...


Searching: 100%|██████████| 10/10 [02:45<00:00, 16.57s/it]


### Παραδείγματα Αποτελεσμάτων

Ας δούμε τα αποτελέσματα για ένα παράδειγμα query για να καταλάβουμε πώς επιδρά η αναζήτηση Bag-of-Words ποιοτικά.

In [20]:
# Show first query and its results
example_query_id = sample_query_ids[0]
example_query = sample_queries[example_query_id]
example_results = bow_results[example_query_id]

print(f"\nExample Query: {example_query}")
print("Top 5 results:")
for i, (doc_id, score) in enumerate(example_results[:5]):
    doc_title = corpus[doc_id]['title']
    doc_text = corpus[doc_id]['text'][:100] + "..." if len(corpus[doc_id]['text']) > 100 else corpus[doc_id]['text']

    print(f"{i+1}. {doc_title} (Score: {score})")
    print(f"   {doc_text}")

    if example_query_id in qrels and doc_id in qrels[example_query_id]:
        print(f"   [RELEVANT: Score {qrels[example_query_id][doc_id]}]")
    print()


Example Query: Do Cholesterol Statin Drugs Cause Breast Cancer?
Top 5 results:
1. Elevated Levels of Cholesterol-Rich Lipid Rafts in Cancer Cells Are Correlated with Apoptosis Sensitivity Induced by Cholesterol-Depleting Agents (Score: 5)
   Lipid rafts/caveolae are membrane platforms for signaling molecules that regulate various cellular f...
   [RELEVANT: Score 2]

2. Statin Use and Breast Cancer Survival: A Nationwide Cohort Study from Finland (Score: 4)
   Recent studies have suggested that statins, an established drug group in the prevention of cardiovas...
   [RELEVANT: Score 2]

3. Statin use after diagnosis of breast cancer and survival: a population-based cohort study. (Score: 4)
   BACKGROUND: Preclinical studies have shown that statins, particularly simvastatin, can prevent growt...
   [RELEVANT: Score 2]

4. Statin use and risk of breast cancer: a meta-analysis of observational studies. (Score: 4)
   Emerging evidence suggests that statins' may decrease the risk of cancers

### Ποσοτική Αξιολόγηση: NDCG & RECALL

Τέλος, ας αξιολογήσουμε ποσοτικά την αναζήτηση Bag-of-Words χρησιμοποιώντας το NDCG και το Recall.


## Υπολογισμός Μετρικών Αξιολόγησης

**Τι κάνει ένας καλός αλγόριθμος ανάκτησης;** Διαισθητικά, πρέπει να κατατάσσει τα πιο σχετικά έγγραφα ψηλότερα από τα λιγότερο σχετικά έγγραφα. Υπάρχουν πολλοί τρόποι με τους οποίους αυτό μπορεί να ποσοτικοποιηθεί.

Θα υπολογίσουμε το Normalized Discounted Cumulative Gain ([nDCG@K](https://en.wikipedia.org/wiki/Discounted_cumulative_gain)) και το [Recall@K](https://en.wikipedia.org/wiki/Precision_and_recall).

### Συνοπτικά: nDCG vs Recall

**nDCG**: Ένα καλό αποτέλεσμα ``nDCG@K`` μας λέει ότι ο αλγόριθμος ανάκτησης κατατάσσει μερικά από τα πιο σχετικά έγγραφα πολύ ψηλά (αλλά μπορεί κάποια σχετικά έγγραφα να μην εμφανίζονται καν στα πρώτα Κ).

**Recall**: Ένα καλό αποτέλεσμα ``Recall@K`` μας λέει ότι μεγάλο ποσοστό των σχετικών εγγράφων κατατάσσονται στα πρώτα Κ αποτελέσματα (αλλά μπορεί να μην βρίσκονται πολύ ψηλά στη λίστα).


## NDCG και Recall: οι λεπτομέρειες των υπολογισμών

### 1. Normalized Discounted Cumulative Gain (nDCG@K)
Το **nDCG@K** είναι μια μετρική αξιολόγησης μιας κατάταξης, που ποσοτικοποιεί πόσο καλά μια μέθοδος αναζήτησης ή ένα ranking μοντέλο ταξινομεί σχετικά έγγραφα στα πρώτα $K$ αποτελέσματα. Λαμβάνει υπόψη και τα δύο:
- **Relevance** των ανακτηθέντων εγγράφων.
- **Position** (τα ψηλότερα στην κατάταξη σχετικά έγγραφα είναι πιο πολύτιμα).

Το nDCG υπολογίζεται ως:

$$
\text{nDCG@K} = \frac{\text{DCG@K}}{\text{IDCG@K}}
$$

όπου:

1. **Discounted Cumulative Gain (DCG@K)**:

$$
\text{DCG@K} = \sum_{i=1}^{K} \frac{\text{rel}_i}{\log_2(i + 1)}
$$

- $ \text{rel}_i $ = relevance score του εγγράφου στη θέση $ i $.
- $ \log_2(i + 1) $ μειώνει τη σημασία των χαμηλότερα στην κατάταξη σχετικών εγγράφων.

2. **Ideal DCG (IDCG@K)**:
- Το **καλύτερο δυνατό DCG**: έγγραφα ταξινομημένα σε ιδανική/τέλεια σειρά: $\text{rel}_1 > \text{rel}_2 > \text{rel}_3$ κ.ο.κ.
- Υπολογίζεται με τον ίδιο τρόπο με το DCG αλλά χρησιμοποιώντας την **ιδανική κατάταξη**.

Έτσι, το **nDCG@K είναι πάντα μεταξύ 0 και 1**, όπου:
- **1.0** σημαίνει τέλεια κατάταξη.
- **Χαμηλότερες τιμές** υποδηλώνουν υποβέλτιστη κατάταξη.

#### Παράδειγμα Υπολογισμού
Ας φανταστούμε ότι κατατάσσουμε **3 έγγραφα** όπου τα relevance scores τους (ground truth) είναι **[3, 2, 1]**.

Εάν το μοντέλο τα κατατάσσει ως **[2, 3, 1]**, υπολογίζουμε:

$$
\text{DCG@3} = \frac{2}{\log_2(1+1)} + \frac{3}{\log_2(2+1)} + \frac{1}{\log_2(3+1)}
$$

$$
= 2 + \frac{3}{1.585} + \frac{1}{2} \approx 3.892
$$

Για το **IDCG@3** (ιδανική κατάταξη [3,2,1]):

$$
\text{IDCG@3} = \frac{3}{\log_2(1+1)} + \frac{2}{\log_2(2+1)} + \frac{1}{\log_2(3+1)}
$$

$$
= 3 + \frac{2}{1.585} + \frac{1}{2} \approx 4.892
$$

$$
\text{nDCG@3} = \frac{3.892}{4.892} \approx 0.796
$$

Αυτό σημαίνει ότι η κατάταξή μας είναι **79.6% τόσο καλή** όσο μια ιδανική κατάταξη.

---

### 2. Recall@K
Το **Recall@K** μετρά πόσα σχετικά έγγραφα ανακτώνται στα πρώτα $K$ αποτελέσματα, σε σύγκριση με όλα τα σχετικά έγγραφα στο dataset.

#### Formula
$$
\text{Recall@K} = \frac{\text{Αριθμός σχετικών εγγράφων στα πρώτα K}}{\text{Συνολικός αριθμός σχετικών εγγράφων στο corpus}}
$$

- Εάν ένα dataset έχει **10 σχετικά έγγραφα** και το σύστημα ανακτά **4 από αυτά στα top 5**, τότε:
$$
  \text{Recall@5} = \frac{4}{10} = 0.4
$$
  που σημαίνει ότι το σύστημα ανακάλυψε **40% όλων των σχετικών εγγράφων** στα top 5 αποτελέσματα.

#### Ερμηνεία
- **Υψηλό Recall@K** → Το μοντέλο βρίσκει περισσότερα σχετικά έγγραφα.
- **Χαμηλό Recall@K** → Πολλά σχετικά έγγραφα λείπουν από τα top αποτελέσματα.

#### Σύγκριση με Precision@K
Το Precision@K είναι μια άλλη μετρική αξιολόγησης που επίσης χρησιμοποιείται συχνά, και ποσοτικοποιεί πόσα από τα πρώτα Κ άρθρα στην κατάταξη είναι σχετικά. Οπότε για το παράδειγμά μας με 10 σχετικά έγγραφα όπου το σύστημα ανακτά 4 από αυτά στα πρώτα 5, τότε
$$
  \text{Precision@5} = \frac{4}{5} = 0.4
$$
Εάν υπάρχουν 2 σχετικά άρθρα και το σύστημα ανακτά και τα 2 στα πρώτα πέντε, τότε έχουμε
$$
  \text{Recall@5} = \frac{2}{2} = 1
$$
$$
  \text{Precision@5} = \frac{2}{5} = 0.4
$$


---

### Πότε να Χρησιμοποιήσουμε nDCG@K vs Recall@K
- Το **nDCG@K** είναι χρήσιμο όταν **η σειρά κατάταξης έχει σημασία**, όπως σε search engines.
- Το **Recall@K** είναι καλύτερο όταν σας ενδιαφέρει να **ανακτήσετε όλα τα σχετικά docs** (π.χ., αναζήτηση νομικών εγγράφων, ή αν χτίζετε ένα RAG).


In [21]:
def calculate_metrics(qrels, results, k_values):
    """Calculate NDCG and Recall metrics for search results."""
    ndcg = {f"NDCG@{k}": 0.0 for k in k_values}
    recall = {f"Recall@{k}": 0.0 for k in k_values}

    for query_id in qrels:
        if query_id not in results:
            continue

        relevant_docs = set(qrels[query_id].keys())
        retrieved_docs = list(results[query_id].keys())

        # Calculate NDCG
        for k in k_values:
            dcg = 0
            idcg = sum((1 / math.log2(i + 2) for i in range(min(k, len(relevant_docs)))))

            for i, doc_id in enumerate(retrieved_docs[:k]):
                if doc_id in relevant_docs:
                    dcg += 1 / math.log2(i + 2)

            if idcg > 0:
                ndcg[f"NDCG@{k}"] += dcg / idcg

        # Calculate Recall
        for k in k_values:
            retrieved_relevant = len(set(retrieved_docs[:k]) & relevant_docs)
            if len(relevant_docs) > 0:
                recall[f"Recall@{k}"] += retrieved_relevant / len(relevant_docs)

    # Average the metrics
    num_queries = len(qrels)
    for k in k_values:
        ndcg[f"NDCG@{k}"] /= num_queries
        recall[f"Recall@{k}"] /= num_queries

    return ndcg, recall

### NDCG & Recall

Στα 10 queries που χρησιμοποιήσαμε.

In [22]:
# Convert results to format expected by metrics function
formatted_results = {
    query_id: {doc_id: score for doc_id, score in results}
    for query_id, results in bow_results.items()
}

# Calculate metrics
ndcg, recall = calculate_metrics(sample_qrels, formatted_results, k_values)

# Display results
print("\nBag of Words Search Metrics:")
for k in k_values:
    print(f"NDCG@{k}: {ndcg[f'NDCG@{k}']:.4f}")
    print(f"Recall@{k}: {recall[f'Recall@{k}']:.4f}")


Bag of Words Search Metrics:
NDCG@5: 0.2693
Recall@5: 0.0362
NDCG@10: 0.1899
Recall@10: 0.0394
NDCG@20: 0.1259
Recall@20: 0.0411
NDCG@100: 0.1078
Recall@100: 0.0862


## Πολυπλοκότητα & Search Indexing

**Η διαδικασία αναζήτησης δεν ήταν αποδοτική** Χρειάστηκε πολύς χρόνος μόνο για να επεξεργαστούμε ``sample_size=10`` queries. Τι θα γινόταν αν θέλαμε να επεξεργαστούμε όλα;

### Το Πρόβλημα: Γραμμική Αναζήτηση (Linear Search)
Χωρίς ευρετήριο (index), η αναζήτηση σε έγγραφα είναι σαν να ψάχνει κανείς σε κάθε σελίδα κάθε βιβλίου σε μια βιβλιοθήκη για να βρούμε αναφορές μιας λέξης. Για κάθε query:
- Πρέπει να κοιτάξουμε κάθε έγγραφο
- Για κάθε έγγραφο, πρέπει να σαρώσουμε όλες τις λέξεις του
- Πολυπλοκότητα χρόνου: O(N × M) όπου N = αριθμός εγγράφων, M = μέσο μήκος εγγράφου

### Η Λύση: Inverted Index
Ένα inverted index είναι σαν ένα ευρετήριο βιβλίου αλλά για ολόκληρη τη συλλογή εγγράφων. Αντί να οργανώνουμε πρώτα κατά έγγραφα, οργανώνουμε κατά λέξεις:

```python
# Χωρίς index (αρχική προσέγγιση):
documents = {
    'doc1': 'the cat sat',
    'doc2': 'the dog ran',
    'doc3': 'cat and dog'
}

# Με inverted index:
index = {
    'the': {'doc1', 'doc2'},
    'cat': {'doc1', 'doc3'},
    'dog': {'doc2', 'doc3'},
    'sat': {'doc1'},
    'ran': {'doc2'},
    'and': {'doc3'}
}
```

### Γιατί είναι Πιο Γρήγορο

**Αρχική Αναζήτηση (Linear)**
Για query "cat dog":
1. Κοιτάμε doc1: σαρώνουμε όλες τις λέξεις → βρίσκουμε 'cat'
2. Κοιτάμε doc2: σαρώνουμε όλες τις λέξεις → βρίσκουμε 'dog'
3. Κοιτάμε doc3: σαρώνουμε όλες τις λέξεις → βρίσκουμε και τα δύο
- Πρέπει να εξετάσουμε ΚΑΘΕ λέξη σε ΚΑΘΕ έγγραφο

**Indexed Search**
Για query "cat dog":
1. Αναζητούμε 'cat' → παίρνουμε αμέσως {doc1, doc3}
2. Αναζητούμε 'dog' → παίρνουμε αμέσως {doc2, doc3}
3. Συνδυάζουμε αποτελέσματα → το doc3 έχει και τις δύο λέξεις
- Εξετάζουμε μόνο τα σχετικά έγγραφα
- Άμεση πρόσβαση σε θέσεις λέξεων

### Σύγκριση Απόδοσης
- **Linear Search**: O(N × M) πράξεις
  - N = αριθμός εγγράφων
  - M = μέσο μήκος εγγράφου
  
- **Indexed Search**: O(k × |posting_list|) πράξεις
  - k = αριθμός όρων query
  - |posting_list| = μέσος αριθμός εγγράφων ανά λέξη
  - Συνήθως |posting_list| << N × M

### Συμβιβάσεις
- **Χώρος**: Το index απαιτεί πρόσθετη αποθήκευση
- **Χρόνος Προεπεξεργασίας**: Πρέπει να χτιστεί το ευρετήριο πριν την αναζήτηση
- **Κόστος Ενημέρωσης**: Η προσθήκη/τροποποίηση εγγράφων απαιτεί ενημερώσεις στο ευρετήριο
- **Όφελος**: Δραματικά ταχύτερος χρόνος αναζήτησης

**Απλό Παράδειγμα**
Με 10.000 έγγραφα που έχουν κατά μέσο όρο 1.000 λέξεις το καθένα:
- Linear search: Πρέπει να εξετάσουμε 10.000.000 λέξεις ανά query
- Indexed search: Μπορεί να εξετάσουμε μόνο μερικές εκατοντάδες document IDs ανά λέξη query

Αυτός είναι ο λόγος που η τροποποιημένη υλοποίηση Bag of Words μας με index μπορεί να χειριστεί αποτελεσματικά το πλήρες dataset, και ο λόγος που αρχικά περιοριστήκαμε σε 10 queries.

In [23]:
class BagOfWordsSearch2:
    def __init__(self, use_stemming: bool = False):
        """Initialize the Bag of Words search system with an inverted index."""
        self.documents = {}  # Document storage
        self.index = {}     # Inverted index: word -> set of doc_ids
        self.use_stemming = use_stemming
        if use_stemming:
            self.stemmer = PorterStemmer()

    def add_document(self, doc_id: str, text: str) -> None:
        """Add a document to the search index."""
        # Store the original document
        self.documents[doc_id] = text

        # Process the document
        tokens = self._preprocess_text(text)

        # Add each token to the inverted index
        for token in tokens:
            if token not in self.index:
                self.index[token] = set()
            self.index[token].add(doc_id)

    def _preprocess_text(self, text: str) -> List[str]:
        """Preprocess text by tokenizing and optionally stemming."""
        tokens = word_tokenize(text.lower())
        tokens = [re.sub(r'[^a-z0-9]', '', token) for token in tokens]
        tokens = [token for token in tokens if token]

        if self.use_stemming:
            tokens = [self.stemmer.stem(token) for token in tokens]

        return tokens

    def search(self, query: str, top_k: Optional[int] = None) -> List[tuple]:
        """Search documents using bag of words matching with inverted index."""
        query_tokens = self._preprocess_text(query)
        doc_scores = {}  # doc_id -> score

        # For each query token, find matching documents using the index
        for token in query_tokens:
            if token in self.index:
                for doc_id in self.index[token]:
                    doc_scores[doc_id] = doc_scores.get(doc_id, 0) + 1

        # Convert to list and sort by score
        results = [(doc_id, score) for doc_id, score in doc_scores.items()]
        results.sort(key=lambda x: x[1], reverse=True)

        if top_k is not None:
            results = results[:top_k]

        return results

# Now let's use it on the full dataset
print("Initializing search engine...")
bow_search = BagOfWordsSearch2(use_stemming=True)

# Add all documents to the index
print("\nBuilding index...")
for doc_id, doc in tqdm(corpus.items(), desc="Indexing documents"):
    full_text = f"{doc['title']} {doc['text']}"
    bow_search.add_document(doc_id, full_text)

# Run all queries
print("\nRunning searches...")
bow_results = {}
for query_id, query_text in tqdm(queries.items(), desc="Processing queries"):
    results = bow_search.search(query_text, top_k=max(k_values))
    bow_results[query_id] = results

# Calculate metrics using the same code as before
formatted_results = {
    query_id: {doc_id: score for doc_id, score in results}
    for query_id, results in bow_results.items()
}

# Calculate metrics
ndcg, recall = calculate_metrics(qrels, formatted_results, k_values)

# Display results
print("\nBag of Words Search Metrics (Full Dataset):")
for k in k_values:
    print(f"NDCG@{k}: {ndcg[f'NDCG@{k}']:.4f}")
    print(f"Recall@{k}: {recall[f'Recall@{k}']:.4f}")

Initializing search engine...

Building index...


Indexing documents: 100%|██████████| 3633/3633 [00:18<00:00, 195.35it/s]



Running searches...


Processing queries: 100%|██████████| 323/323 [00:00<00:00, 1536.88it/s]


Bag of Words Search Metrics (Full Dataset):
NDCG@5: 0.2641
Recall@5: 0.0956
NDCG@10: 0.2339
Recall@10: 0.1127
NDCG@20: 0.2109
Recall@20: 0.1326
NDCG@100: 0.2042
Recall@100: 0.1970


## Δημιουργία Sparse Index με Pyserini

Στις εντολές παραπάνω χτίσαμε μόνοι μας το sparse index και υπολογίσαμε το Bag-of-Words και το TF-IDF. Στην πράξη, ιδιαίτερα για μεγαλύτερες συλλογές κειμένων, χρησιμοποιούμε βιβλιοθήκες που βελτιστοποιούν αυτή τη διαδικασία.

Για να επιδείξουμε μία από αυτές τις βιβλιοθήκες, κατασκευάζουμε τώρα ένα sparse index χρησιμοποιώντας μια Python βιβλιοθήκη που λέγεται [Pyserini](https://github.com/castorini/pyserini). Για να λειτουργήσει η Pyserini, απαιτεί μια συγκεκριμένη δομή δεδομένων που λέγεται JSON, η οποία είναι ένας τρόπος αποθήκευσης λεξικών. Για παράδειγμα,
```
{
  "όνομα": "Κωνσταντίνος",
  "ηλικία": 48,
  "πόλεις": ["Αθήνα", "Ώστιν"]
}
```
Θα διαμορφώσουμε τα δεδομένα μας ώστε να ταιριάζουν σε αυτή τη μορφή.


In [24]:
#making a temporary directory to store the index
tmp_dir = tempfile.mkdtemp()

# Prepare the corpus for Pyserini indexing
corpus_path = os.path.join(tmp_dir, 'corpus.jsonl')
with open(corpus_path, 'w') as f:
    for doc_id, doc in corpus.items():
        json.dump({'id': doc_id, 'contents': doc['text']}, f)
        f.write('\n')

sparse_index_dir = os.path.join(tmp_dir, 'index')

# αυτή η εντολή χτίζει το sparse index με την Pyserini
os.system(f"python -m pyserini.index -collection JsonCollection -generator DefaultLuceneDocumentGenerator -threads 1 -input {tmp_dir} -index {sparse_index_dir} -storePositions -storeDocvectors -storeRaw")

0

### Ανάκτηση

Θα ανακτήσουμε τώρα έγγραφα με την Pyserini. Έχουμε ήδη δει πώς η Pyserini χτίζει το ευρετήριο (inverted index). Η Pyserini διαθέτει επίσης βιβλιοθήκες για την ανάκτηση. Ενώ στο πρώτο μέρος του notebook υλοποιήσαμε μόνοι μας το TF-IDF, εδώ χρησιμοποιούμε τη βιβλιοθήκη ``LuceneSearcher`` της Pyserini.

Το `LuceneSearcher` χρησιμοποιεί BM25, μια μέθοδο βαθμολόγησης πολύ συγγενική με το TF-IDF, για να κατατάξει τα έγγραφα με βάση το πόσο καλά ταιριάζουν με το query. Μόλις αναγνωριστούν τα σχετικά έγγραφα, ο searcher τα ανακτά σε σειρά κατάταξης, επιστρέφοντας τα top-scoring αποτελέσματα.

Στη συνέχεια εξηγούμε το BM25.


**BM25 (Best Match 25)**

Το BM25 είναι μια προηγμένη ranking function που βελτιώνει το TF-IDF αντιμετωπίζοντας δύο βασικούς περιορισμούς:
1. Term frequency saturation (φθίνουσες αποδόσεις)
2. Document length normalization

BM25 score για ένα έγγραφο D και query Q:
$$
\text{score}(D,Q) = \sum_{i=1}^n \text{IDF}(q_i) \cdot \frac{f(q_i,D) \cdot (k_1 + 1)}{f(q_i,D) + k_1 \cdot (1 - b + b \cdot \frac{|D|}{\text{avgdl}})}
$$

Όπου:
- $f(q_i,D)$: Term frequency της λέξης qi στο έγγραφο D
- $|D|$: Μήκος του εγγράφου D
- avgdl: Μέσο μήκος εγγράφου
- $k_1$: Παράμετρος term frequency saturation (τυπικά 1.2-2.0)
- $b$: Παράμετρος length normalization (τυπικά 0.75)

**Βασικές Βελτιώσεις σε σχέση με TF-IDF**

1. Term Frequency Saturation
- TF-IDF: Γραμμική αύξηση με term frequency
- BM25: Φθίνουσες αποδόσεις καθώς αυξάνεται το term frequency
- Παράδειγμα: Εάν το "neural" εμφανίζεται 10 φορές vs 100 φορές, η διαφορά στο score δεν είναι 10x

2. Document Length Normalization
- TF-IDF: Δεν λαμβάνει υπόψη το μήκος εγγράφου
- BM25: Κανονικοποιεί με βάση το μήκος εγγράφου
- Τα μακρύτερα έγγραφα δεν παίρνουν αυτόματα υψηλότερα scores

3. Απλό Παράδειγμα
Query: "neural networks"
- Σύντομο paper με 2 αναφορές: μπορεί να πάρει υψηλότερο score
- Μακρύ paper με 3 αναφορές: μπορεί να πάρει χαμηλότερο score
- Το BM25 λαμβάνει υπόψη:
  - Μήκος εγγράφου vs. μέσο μήκος
  - Φθίνοντα value επαναλαμβανόμενων όρων
  - Σπανιότητα όρων στη συλλογή

Αυτός είναι ο λόγος που το BM25 συχνά ξεπερνά απλούστερες μεθόδους ranking.

In [25]:
sparse_searcher = LuceneSearcher(sparse_index_dir)

results = {}
for query_id, query in tqdm(queries.items(), total=len(queries), desc="Retrieving"):
    hits = sparse_searcher.search(query, k=100)

    #we will store the results in a dictionary in the form of {doc_id: score} pairs, similar to how it is stored in the ground-truth qrels object.
    results[query_id] = {hit.docid: hit.score for hit in hits}

Retrieving: 100%|██████████| 323/323 [00:03<00:00, 104.86it/s]


Θα υπολογίσουμε nDCG@10 και Recall@100. Μπορούμε να τροποποιήσουμε το `k_values` για να πάρουμε διαφορετικές τιμές του K (όσο είναι <= 100, δηλαδή όσος είναι ο αριθμός εγγράφων που ανακτούμε στο `sparse_searcher.search()` παραπάνω).


In [26]:
# Configuration
k_values = [5, 10, 20, 100]

# Calculate metrics for TF-IDF search
tfidf_ndcg, tfidf_recall = calculate_metrics(qrels, results, k_values)

# Display results
print("\nBM25 Search Metrics:")
for k in k_values:
    print(f"NDCG@{k}: {tfidf_ndcg[f'NDCG@{k}']:.4f}")
    print(f"Recall@{k}: {tfidf_recall[f'Recall@{k}']:.4f}")


BM25 Search Metrics:
NDCG@5: 0.3505
Recall@5: 0.1198
NDCG@10: 0.3131
Recall@10: 0.1504
NDCG@20: 0.2805
Recall@20: 0.1756
NDCG@100: 0.2578
Recall@100: 0.2415


### Σύγκριση Απόδοσης: Bag of Words vs TF-IDF/BM25

Τα metrics δείχνουν ότι η αναζήτηση TF-IDF/BM25 ξεπερνά τη βασική προσέγγιση Bag of Words.

**NDCG (Normalized Discounted Cumulative Gain)**
- **Top 5 Αποτελέσματα**: Το TF-IDF επιτυγχάνει nDCG@5 ίσο με 0.3505, περίπου 38% καλύτερο από το Bag of Words (0.2541)
- **Top 10 Αποτελέσματα**: Το πλεονέκτημα συνεχίζεται, με nDCG@10 ίσο με 0.3131 έναντι 0.2289, διατηρώντας καλύτερη ποιότητα κατάταξης
- **Top 20 Αποτελέσματα**: Το TF-IDF δείχνει nDCG@20 ίσο με 0.2805, έναντι 0.2082 για το BoW
- **Top 100 Αποτελέσματα**: Ακόμα και σε μεγαλύτερα σύνολα αποτελεσμάτων, το TF-IDF διατηρεί το προβάδισμα (0.2578 έναντι 0.2010)

**Recall (Ικανότητα Εύρεσης Σχετικών Εγγράφων)**
- **Top 5 Αποτελέσματα**: Το TF-IDF βρίσκει ελαφρώς περισσότερα σχετικά έγγραφα (11.98% έναντι 9.19%)
- **Top 10 Αποτελέσματα**: Η βελτίωση συνεχίζεται, με 15.04% recall για το TF-IDF έναντι 11.24% για το Bag of Words
- **Top 20 Αποτελέσματα**: Το TF-IDF επιτυγχάνει 17.56% recall έναντι 13.32% για το BoW
- **Top 100 Αποτελέσματα**: Σε μεγαλύτερα σύνολα αποτελεσμάτων, το TF-IDF βρίσκει περίπου 25% περισσότερα σχετικά έγγραφα (24.15% έναντι 19.18%)

**Τι σημαίνουν αυτοί οι αριθμοί:** Με την αξιολόγηση του πλήρους dataset, μπορούμε να δούμε ότι, ενώ το TF-IDF/BM25 εξακολουθεί να ξεπερνά το Bag of Words, η διαφορά είναι πιο λεπτή:

1. **Ποιότητα Κατάταξης**: Το TF-IDF παρέχει συστηματικά καλύτερη κατάταξη (υψηλότερο nDCG σε όλα τα επίπεδα), υποδηλώνοντας ότι είναι καλύτερο στο να τοποθετεί τα πιο σχετικά έγγραφα πρώτα
2. **Εύρεση Εγγράφων**: Και οι δύο μέθοδοι μπορούν να βρουν σχετικά έγγραφα, αλλά το TF-IDF είναι πιο αποτελεσματικό, με 25–30% βελτίωση στο recall
3. **Συνέπεια**: Το πλεονέκτημα του TF-IDF διατηρείται σε διαφορετικά μεγέθη συνόλων αποτελεσμάτων, δείχνοντας σταθερή απόδοση


In [27]:
#deleting temporary files
shutil.rmtree(tmp_dir)

### Άσκηση -- κατασκευάστε το δικό σας Corpus

Υπάρχουν ουσιαστικά άπειροι πόροι στο διαδίκτυο για να κατασκευάσετε το δικό σας σώμα κειμένων και να υλοποιήσετε τις παραπάνω ιδέες/μεθόδους. Σας προτείνω έναν εύκολο τρόπο: να κατεβάσετε βιβλία από το project [gutenbergpy](https://github.com/raduangelescu/gutenbergpy), για παράδειγμα 4 βιβλία του Mark Twain (τα IDs των βιβλίων παρέχονται):

  * "Roughing It": 3178,
  * "Life on the Mississippi": 1937,
  * "The Adventures of Tom Sawyer": 74,
  * "Adventures of Huckleberry Finn": 1861

Με αυτό, ή με άλλο σώμα κειμένων, κατασκευάστε ένα corpus σε παρόμοια μορφή με το παραπάνω παράδειγμα. Θα χρειαστεί να χωρίσετε τα βιβλία σε κείμενα ('documents'), για παράδειγμα σε κεφάλαια, παραγράφους ή προτάσεις. Γενικότερα, για κάθε σώμα κειμένων πρέπει να σκεφτούμε από την αρχή ποια μέθοδος διαίρεσης ταιριάζει στο περιεχόμενο.

Προτείνω να χρησιμοποιήσετε τους ακόλουθους ορισμούς του τι είναι ένα 'document' και να τους δοκιμάσετε όλους:

  * Κάθε κεφάλαιο είναι ένα document (`CHAPTER\n` ως delimiter)
  * Κάθε παράγραφος είναι ένα document (`\n\n` ως delimiter)
  * Κάθε πρόταση είναι ένα document (`.` ως delimiter)

Θα σας είναι επίσης χρήσιμες οι βιβλιοθήκες επεξεργασίας γλώσσας που χρησιμοποιήσαμε, όπως η βιβλιοθήκη ``NLTK`` και η βιβλιοθήκη ``re``.

Δημιουργήστε ένα index και στη συνέχεια γράψτε μερικά queries. Ανακτήστε τα 'documents' σας χρησιμοποιώντας αυτά τα queries. Απεικονίστε τα ανακτηθέντα έγγραφα για καθεμία από τις παραπάνω διαμορφώσεις.


## Χρήσιμες Εντολές

Σας προτείνουμε κάποιες χρήσιμες εντολές που μπορείτε να χρησιμοποιήσετε.


In [28]:
!pip install gutenbergpy

  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 11.5 MB/s  0:00:00
  Created wheel for httpsproxy-urllib2: filename=httpsproxy_urllib2-1.0-py3-none-any.whl size=29292 sha256=08042795e3be27e82fcc6f0231a092a5dcaf56aa9e5c9050a349f0f1f7f513cc
  Stored in directory: /root/.cache/pip/wheels/c6/c5/98/dd27835c8319362f1cef163409f43df38d54f084500fad428c
Successfully built httpsproxy-urllib2
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3/3 [gutenbergpy]


In [29]:
# List of book titles and their corresponding Project Gutenberg IDs
books = {
    "Roughing It": 3177,
    "Life on the Mississippi": 245,
    "The Adventures of Tom Sawyer": 74,
    "Adventures of Huckleberry Finn": 76
}

In [30]:
import gutenbergpy.textget
import re

def clean_text(text):
    """Clean up the text by removing headers, footers, chapter titles, etc."""
    # Convert bytes to string if needed
    if isinstance(text, bytes):
        text = text.decode('utf-8')

    # Remove Project Gutenberg headers and footers
    text = gutenbergpy.textget.strip_headers(text.encode('utf-8')).decode('utf-8')

    # Remove image references and file sizes
    text = re.sub(r'\([0-9]+K\)', '', text)
    text = re.sub(r'\.jpg', '', text)

    # Remove table of contents and metadata
    text = re.sub(r'^.*?CONTENTS.*?\n.*?\n', '', text, flags=re.DOTALL)
    text = re.sub(r'^.*?TABLE OF CONTENTS.*?\n.*?\n', '', text, flags=re.DOTALL)
    text = re.sub(r'^.*?By.*?\n.*?\n', '', text, flags=re.DOTALL)
    text = re.sub(r'^.*?PREFATORY.*?\n.*?\n', '', text, flags=re.DOTALL)
    text = re.sub(r'^.*?INTRODUCTION.*?\n.*?\n', '', text, flags=re.DOTALL)

    # Remove chapter headings but keep the content
    text = re.sub(r'CHAPTER [IVXLC]+\.*\s*\n', '', text)
    text = re.sub(r'ILLUSTRATIONS?\s*\n+(?:[^\n]+\n+)*', '', text)

    # Remove extra whitespace but preserve paragraph breaks
    text = re.sub(r'[ \t]+', ' ', text)
    text = re.sub(r'\n{3,}', '\n\n', text)

    return text.strip()

def split_into_paragraphs(text, min_length=50):
    """Split text into meaningful paragraphs."""
    # Split on double newlines
    paragraphs = re.split(r'\n\s*\n', text)

    # Clean up each paragraph
    paragraphs = [p.strip() for p in paragraphs]

    # Remove short paragraphs and empty lines
    paragraphs = [p for p in paragraphs if len(p) >= min_length]

    # Remove paragraphs that are just lists or titles
    paragraphs = [p for p in paragraphs if not re.match(r'^[A-Z][a-z]*\s*$', p)]
    paragraphs = [p for p in paragraphs if not re.match(r'^[IVXLC]+\.', p)]

    return paragraphs

def download_and_process_books(books):
    corpus = {}
    doc_id = 0

    for title, book_id in books.items():
        print(f"\nProcessing {title}...")

        # Download the book
        raw_book = gutenbergpy.textget.get_text_by_id(book_id)

        # Clean the text
        clean_book = clean_text(raw_book)
        print(f"Text length after cleaning: {len(clean_book)}")

        # Split into paragraphs
        paragraphs = split_into_paragraphs(clean_book)

        # Add to corpus with metadata
        for para in paragraphs:
            corpus[f"doc_{doc_id}"] = {
                "text": para,
                "book": title,
                "book_id": book_id
            }
            doc_id += 1

        print(f"Added {len(paragraphs)} paragraphs from {title}")

    return corpus

# Orocess the corpus
mt_corpus = download_and_process_books(books)




Processing Roughing It...
Text length after cleaning: 908225
Added 1505 paragraphs from Roughing It

Processing Life on the Mississippi...
Text length after cleaning: 802471
Added 1551 paragraphs from Life on the Mississippi

Processing The Adventures of Tom Sawyer...
Text length after cleaning: 374319
Added 1205 paragraphs from The Adventures of Tom Sawyer

Processing Adventures of Huckleberry Finn...
Text length after cleaning: 564747
Added 1653 paragraphs from Adventures of Huckleberry Finn


In [31]:
# Create the index
tmp_dir = tempfile.mkdtemp()
corpus_path = os.path.join(tmp_dir, 'mt_corpus.jsonl')
with open(corpus_path, 'w') as f:
    for doc_id, doc in mt_corpus.items():
        json.dump({'id': doc_id, 'contents': doc['text']}, f)
        f.write('\n')

index_dir = os.path.join(tmp_dir, 'index')
os.system(f"python -m pyserini.index -collection JsonCollection -generator DefaultLuceneDocumentGenerator -threads 1 -input {tmp_dir} -index {index_dir} -storePositions -storeDocvectors -storeRaw")
searcher = LuceneSearcher(index_dir)


In [32]:
# code for testing a query
def display_search_results(query, searcher, corpus, k=5):
    """Display search results with better formatting."""
    print(f"\nQuery: {query}")
    print("-" * 80)

    hits = searcher.search(query, k=k)
    if not hits:
        print("No results found.")
        return

    for i, hit in enumerate(hits, 1):
        doc = corpus[hit.docid]
        print(f"\nResult {i} (Score: {hit.score:.3f}) from '{doc['book']}':")
        # Show a shorter preview for better readability
        preview = doc['text'][:200] + "..." if len(doc['text']) > 200 else doc['text']
        print(preview)
        print("-" * 80)
